atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.timelock.paxos;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.common.remoting.ServiceNotAvailableException;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosProposer;
+import com.palantir.paxos.PaxosQuorumChecker;
+import com.palantir.paxos.PaxosResponse;
+import com.palantir.paxos.PaxosRoundFailureException;
+import com.palantir.paxos.PaxosValue;
+import com.palantir.timestamp.MultipleRunningTimestampServiceError;
+import com.palantir.timestamp.TimestampBoundStore;
+
+public class PaxosTimestampBoundStore implements TimestampBoundStore {
+    private static final Logger log = LoggerFactory.getLogger(PaxosTimestampBoundStore.class);
+
+    private final PaxosProposer proposer;
+    private final PaxosLearner knowledge;
+
+    private final List<PaxosAcceptor> acceptors;
+    private final List<PaxosLearner> learners;
+    private final long randomWaitBeforeProposalMs;
+
+    @GuardedBy("this")
+    private SequenceAndBound agreedState;
+
+    private final ExecutorService executor = PTExecutors.newCachedThreadPool(PTExecutors.newNamedThreadFactory(true));
+
+    public PaxosTimestampBoundStore(PaxosProposer proposer,
+            PaxosLearner knowledge,
+            List<PaxosAcceptor> acceptors,
+            List<PaxosLearner> learners,
+            long randomWaitBeforeProposalMs) {
+        this.proposer = proposer;
+        this.knowledge = knowledge;
+        this.acceptors = acceptors;
+        this.learners = learners;
+        this.randomWaitBeforeProposalMs = randomWaitBeforeProposalMs;
+    }
+
+    @Override
+    public synchronized long getUpperLimit() {
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(acceptors),
+                a -> ImmutablePaxosLong.of(a.getLatestSequencePreparedOrAccepted()),
```

Comment:
Hmm, it didn't. I think it's not supported yet; there's an [issue](https://github.com/checkstyle/checkstyle/issues/3009) on this.

Anyway changed to `acceptor`, which is clearer.